### PR TITLE
`eth-json-rpc-provider` migration - B2: Replace typescript, jest config files.

### PR DESCRIPTION
**Pull request by MajorLift**
## Explanation

This PR implements the following incremental step in the process for migrating `eth-json-rpc-provider` into the core monorepo:

***

#### 2. Replace config files.
  - [x] Update `tsconfig*.json`, `typedoc.json`, `jest.config.js` to extend from corresponding files in root. Copy contents of corresponding files in other non-root packages.
  - [x] Keep TypeScript compiler flags. 
  - [x] Add tsconfig reference paths for non-root packages that are upstream dependencies of the migration target.
  - [x] Keep Jest coverage threshold values. 
  - [x] Add `deepmerge` as a devDependency.

***

See https://github.com/MetaMask/core/issues/1551#issuecomment-1745665740 for an outline of the entire process.

## References

- Contributes to #1685
- Contributes to #1551

## Changelog

N/A

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

### CODE REVIEW

**Review by mcmire — APPROVED** [review]
Looks good!

File: merged-packages/eth-json-rpc-provider/package.json [hunk 1]
```diff
@@ -35,6 +35,15 @@
     "@types/jest": "^27.4.1",
     "depcheck": "^1.4.3",
     "deepmerge": "^4.2.2",
+    "eslint": "^8.44.0",
```
Comment (mcmire) [inline merged-packages/eth-json-rpc-provider/package.json:38]: Must have missed this the first time. Do we need these here? I think these are already in `package.json`.

Comment (MajorLift) [reply merged-packages/eth-json-rpc-provider/package.json:38]: Good call! These are removed in https://github.com/MetaMask/core/pull/1766, but I forgot that I shouldn't merge that one first.

File: merged-packages/eth-json-rpc-provider/package.json [hunk 2]
```diff
@@ -35,6 +35,15 @@
...
+    "eslint-config-prettier": "^8.5.0",
+    "eslint-import-resolver-typescript": "^2.5.0",
+    "eslint-plugin-import": "^2.26.0",
+    "eslint-plugin-jest": "^27.1.5",
+    "eslint-plugin-jsdoc": "^39.3.3",
+    "eslint-plugin-n": "^15.7.0",
+    "eslint-plugin-prettier": "^4.2.1",
+    "eslint-plugin-promise": "^6.1.1",
```
Comment (MajorLift) [inline merged-packages/eth-json-rpc-provider/package.json:46]:
```suggestion
```

File: merged-packages/eth-json-rpc-provider/package.json (hunk 1 above)
Comment (MajorLift) [reply merged-packages/eth-json-rpc-provider/package.json]: Fixed!

**Review by mcmire — APPROVED** [review]

**Pull request merged**